src/bin_editor/Oni2_editor.re

```diff
@@ -507,11 +507,14 @@ switch (eff) {
 
     // Add a quit handler, so that regardless of how we quit -
     // we have the opportunity to clean up
-    Revery.App.onBeforeQuit(app, () =>
+    Revery.App.onBeforeQuit(app, _code => {
       if (!currentState^.isQuitting) {
-        dispatch(Model.Actions.Quit(true));
-      }
-    )
+        dispatch(Model.Actions.ReallyQuitting);
+      };
+      // We perform asynchronous cleanup, and
+      // exit(0) ourselves when that's complete
+      Revery.App.PreventQuit
```

Comment:
Thanks for working through this! The lifecycle-management pieces are tricky - I really like the way you investigated it by evaluating the different possibilities in #3270  and then implementing the changes in Revery to support this.